music21/stream/tests.py

```diff
@@ -7034,19 +7035,31 @@ def testTransposeByPitchA(self):
         p1.repeatAppend(note.Note('C4'), 4)
         p1.insert(0, i1)
         p1.insert(2, i2)
+        p1.makeMeasures(inPlace=True)
         p2 = Part()
         p2.repeatAppend(note.Note('C4'), 4)
         p2.insert(0, i2)
+        p2.makeMeasures(inPlace=True)
         s = Score()
         s.insert(0, p1)
         s.insert(0, p2)
 
-        self.assertEqual([str(p) for p in p1.pitches], ['C4', 'C4', 'C4', 'C4'])
-        test = p1._transposeByInstrument(inPlace=False)
-        self.assertEqual([str(p) for p in test.pitches], ['F3', 'F3', 'B-3', 'B-3'])
+        def simulateRecursing(myStream, reverse=False) -> List[pitch.Pitch]:
+            '''
+            _transposeByInstrument() doesn't recurse at the moment,
+            so we have to dip into each measure. Not exposed publicly anyway.
+            '''
+            demoPitches = []
+            for m in myStream[Measure]:
+                post = m._transposeByInstrument(inPlace=False, reverse=reverse)
+                for n in post.notes:
+                    demoPitches.extend(n.pitches)
+            return demoPitches
 
-        test = p1._transposeByInstrument(inPlace=False, reverse=True)
-        self.assertEqual([str(p) for p in test.pitches], ['G4', 'G4', 'D4', 'D4'])
+        self.assertEqual([str(p) for p in p1.pitches], ['C4', 'C4', 'C4', 'C4'])
+        self.assertEqual([str(p) for p in simulateRecursing(p1)], ['F3', 'F3', 'B-3', 'B-3'])
+        self.assertEqual([str(p) for p in simulateRecursing(p1, reverse=True)],
```

Comment:
might it be better to make transposeByInstrument recurse?  I thought it was supposed to, or is it only the private method that doesn't?  Calling private methods on anything but `self` is slightly frowned upon, since it doesn't give flexibility to change internal behavior in the futre.

Reply:
`_transposeByInstrument` essentially gets instruments and then transposes. It doesn't recurse. Tranposing recurses, sure, but the instrument search does not recurse. The instrument search is done with `extendDurationsAndGetBoundaries('Instrument')`, which is marked for deprecation, so maybe it's time to rewrite the instrument search and then drop the recursing from toSoundingPitch/toWrittenPitch, which will simplify them even further.

Another benefit of doing that is that it will prevent edge cases from there being superfluous instruments at multiple levels. Bb Clarinet in part and Bb Clarinet in measure == Clarinet in Ab?

Another another benefit is that the helper method to treatAtSoundingPitch won't have to be called on non-self objects and so it can stay private.

Reply:
This probably means I have to restore the special-case for part-like streams. Calling a score should not search downward to find all part instruments and transpose the entire score-slice by that value. Calling a part should search downward to find all measure and voice instruments and tranpsose the entire part-slice by that value.

Reply:
Ended up restoring the special-case for has part-like streams.